From d7274539b81d09684eb4063441fa7d780089ecab Mon Sep 17 00:00:00 2001
From: consts <grudnevkv@gmail.com>
Date: Fri, 2 Mar 2018 07:03:37 +0000
Subject: [PATCH 1/3] Workaround for MX25 chips

TEST: In-system programming a ThinkPad X200 using a clip and
pico-serprog works now. It just doesn't without this hack.

Chip: MX25L6405D

Tested-by: Riku Viitanen <riku.viitanen@protonmail.com>
Change-Id: I43a306b67862b59c1dcd02729e189f3bf73f481b
---
 cli_classic.c        |  5 +++++
 include/programmer.h |  2 ++
 spi.c                | 11 ++++++++++-
 3 files changed, 17 insertions(+), 1 deletion(-)

diff --git a/cli_classic.c b/cli_classic.c
index 7670e79d..b4636ce3 100644
--- a/cli_classic.c
+++ b/cli_classic.c
@@ -72,6 +72,7 @@ static void cli_classic_usage(const char *name)
 	       " -o | --output <logfile>            log output to <logfile>\n"
 	       "      --flash-contents <ref-file>   assume flash contents to be <ref-file>\n"
 	       " -L | --list-supported              print supported devices\n"
+	       " -m | --workaround-mx               keep flash busy before sending command\n"
 #if CONFIG_PRINT_WIKI == 1
 	       " -z | --list-supported-wiki         print supported devices in wiki syntax\n"
 #endif
@@ -274,6 +275,7 @@ int flashprog_classic_main(int argc, char *argv[])
 		{"version",		0, NULL, 'R'},
 		{"output",		1, NULL, 'o'},
 		{"progress",		0, NULL, OPTION_PROGRESS},
+		{"workaround-mx",	0, NULL, 'm'},
 		{NULL,			0, NULL, 0},
 	};
 
@@ -401,6 +403,9 @@ int flashprog_classic_main(int argc, char *argv[])
 			cli_classic_usage(argv[0]);
 			exit(0);
 			break;
+		case 'm': /* --workaround-mx */
+			workaround_mx = 1;
+			break;
 		case OPTION_PROGRESS:
 			show_progress = true;
 			break;
diff --git a/include/programmer.h b/include/programmer.h
index f43cb6f5..1f2c88c4 100644
--- a/include/programmer.h
+++ b/include/programmer.h
@@ -388,6 +388,8 @@ enum ich_chipset {
 	CHIPSET_PANTHER_LAKE,
 };
 
+extern int workaround_mx; /* workaround for MX25* chips, makes flash operations more reliable, less failures */
+
 /* ichspi.c */
 #if CONFIG_INTERNAL == 1
 int ich7_init_spi(void *spibar, enum ich_chipset);
diff --git a/spi.c b/spi.c
index 2ad84c09..6089f8b7 100644
--- a/spi.c
+++ b/spi.c
@@ -49,14 +49,23 @@ static int spi_send_wrapped_command(
 	return mst->multicommand(mst, cmd);
 }
 
+int workaround_mx; /* Make operations with MX25* chips more reliable */
+
 int spi_send_command(const struct flashctx *flash, unsigned int writecnt,
 		     unsigned int readcnt, const unsigned char *writearr,
 		     unsigned char *readarr)
 {
-	if (spi_current_io_mode(flash) != SINGLE_IO_1_1_1)
+	if (spi_current_io_mode(flash) != SINGLE_IO_1_1_1) {
 		return spi_send_wrapped_command(flash->mst.spi, spi_current_io_mode(flash),
 						writecnt, readcnt, writearr, readarr);
+	} else if (workaround_mx) {
+		const unsigned char cmd[JEDEC_READ_OUTSIZE] = {JEDEC_READ, 0, 0, 0};
+		unsigned char buf[256];
+		/* keep flash busy for some time, keep CS warm before sending actual command */
+		flash->mst.spi->command(flash->mst.spi, sizeof(cmd), sizeof(buf), cmd, buf);
+	}
 
+	/* actual command */
 	return flash->mst.spi->command(flash->mst.spi, writecnt, readcnt, writearr, readarr);
 }
 
-- 
2.47.3

